### ISSUE TITLE
Custom CSS

### ISSUE THREAD

**Issue by carlobeltrame** [hitobito/hitobito#1817]
Als hitobito-Administrator*in
möchte ich z.B. die Farben von hitobito ohne Deployment anpassen können,
um Testumgebungen von Produktionsumgebungen zu unterscheiden und viele weitere Use Cases zu ermöglichen

Dazu würde sich eine Gruppeneinstellung auf dem Dachverband eignen, mit der man entweder nur die Hintergrundfarbe der Instanz setzen kann, oder gleich ein Freitextfeld mit dem man beliebiges CSS auf der Seite platzieren kann. Das CSS würde dann in einem `<style>` Block im Header der Seite ausgegeben.

Was meint die Community dazu?
Beteiligt an den [bisherigen Bemühungen](https://github.com/hitobito/hitobito/pull/842) und [Diskussionen](https://github.com/hitobito/hitobito/issues/643) zu Testumgebungs-Hintergrundfarbe waren @simfeld @kronn @Michael-Schaer @maede97 @btemperli.

Gehen wir zu weit wenn wir beliebiges CSS erlauben? Aus Erfahrungen aus einem anderen Projekt wo das geht habe ich gelernt dass es manchmal ein Nachteil sein kann wenn
* das CSS nicht in Git versioniert ist
* eine grössere Anzahl an Leuten dieses CSS bearbeiten können
* CSS-Änderungen bei der Suche nach einem konkreten Bug zu unerwartetem Verhalten führen

Die Flexibilität des user-definierbaren CSS ist in dem anderen Projekt aber gleichzeitig ein riesiger Vorteil.

# ToDo, nachdem die Diskussion ein Ergebnis hat

- [ ] Diskussionsergebnisse zusammenfassen
- [ ] Issue für Umsetzung erstellen
- [ ] ggf. Mockup erstellen
- [ ] Tech-Spec ergänzen
- [ ] ToDo für Umsetzung definieren

**Comment by maede97** [comment]
Ich finde deine Überlegungen schon mal sehr gut, Carlo!
Meiner Meinung nach sollte auf einer Produktiv-Instanz eigentlich alles eingestellt sein und keine Änderungen brauchen (ansonsten kann man auch gut alles ins Git einchecken und versionieren).
Mein Vorschlag wäre (keine Ahnung, ob & wie das machbar wäre), nur auf der Integrationsinstanzen ein "Custom-CSS" Feld hinzuzufügen? So könnte die Farbe für die Integration geändert werden und auf der Produktiv-Instanz gibt es keine Probleme.

**Comment by carlobeltrame** [comment]
Das geht eben nicht, ist dasselbe Problem wie schon bei der einfacheren Variante wo nur die Hintergrundfarbe im SCSS abgeändert werden sollte. Exakt derselbe Container wird auf Integration und Produktion deployed, und auch mit grösstenteils gleichen oder gleichwertigen Umgebungsvariabeln (RAILS_ENV=prod und so weiter). Und sobald wir beginnen, je nach hostnamen anders zu reagieren und daher die hostnamen im Code hardcoden wirds unschön finde ich.

Aber wenn das Feld da wäre könntet ihr natürlich für euch entscheiden, es auf Produktion einfach nicht zu benutzen. Gruppeneinstellungen können nur Personen mit sehr hoher Berechtigung bearbeiten, da wäre der Kreis der Personen vielleicht klein genug dass man das unter sich abmachen könnte.

**Comment by carlobeltrame** [comment]
Ich stimme dir absolut zu wenn du sagst dass das Design der Produktionsumgebung in git viel besser aufgehoben ist als in so einem Feld.

**Comment by nchiapol** [comment]
Ich unterstütze ein solches Feld ebenfalls. Das scheint mir einfach umsetzbar und hinreichend flexibel.

Für das Produktiv-System denke ich auch, dass es reicht, die Admins zu instruieren das sie das Feld leer lassen müssen. Wir sollten unabsichtliche Fehler verhindern, aber wir müssen die Admins nicht vor sich selbst schützen.

**Comment by kronn** [comment]
Für mich wäre ein solches Feld okay, wenn es etwas versteckt ist. Durch die Möglichkeit, eigene Logos für jede Ebenen hochzuladen habe ich gelernt, dass das dann auch genutzt wird. Wenn das Ziel eine Unterscheidung von produktiver Instanz und Integrationsumgebung zu erreichen sollten hier eine zentrale Einstellung, die dann auch nur für admins sichtbar ist, ausreichen. Durch die Positionierung in den Einstellungen (Gruppeneinstellungen nur auf dem Dachverband oder Einstellungen wie Labelformate) ist es sicherlich genug abgeschottet, dass man zweimal nachdenkt. Damit ist dann auch der "Adminselbstschutz" :wink: gegeben. 

Wenn man die Gruppen direkt im Code mit dieser Einstellung versieht (`self.custom_css = true`), könnte der SBV auch den verschiedenen grossen Gruppierungen leicht andere Farben geben. Weiterhin könnte jeder Kunde in der SaaS-Instanz von hitobito.com seine eigenes Design ~~kaputtmachen~~ _verbessern_

Vermutlich sollte aber eine gut sichtbare Warnung vorhanden sein, da man sich leicht die Benutzbarkeit entfernen kann.

**Comment by ThomasEllenberger** [comment]
Da diese Diskussion bereits älter ist und von unserer Seite her aktuell kein Plan existiert dies umzusetzen, schliesse ich dieses Issue vorerst, um die Anzahl offener Issues möglichst übersichtlich zu halten.
Kann bei Bedarf wieder geöffnet werden.
